### PR TITLE
Implement adding and editing widgets

### PR DESCRIPTION
**Pull request by noisecapella**
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots 
  - [x] Mobile width screenshots 
  - [x] tag @ferdi or @pdpinch for review  
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Part of #1508 

#### What's this PR do?
Implements adding and editing widgets, similar to how it's done in the video. The textboxes look different and preview is not implemented in this PR.

#### How should this be manually tested?
Click the gear icon then click 'manage widgets'. You should see '+ Add widget'. Click that and go through the dialog. If you click cancel at any point the dialog should close and nothing should change on the widget list.

If you fill out the items and click 'Create widget' it should add the widget to the widget list. Note that the widget is not saved to the server yet. Click 'Done' and you should see the widget list refresh with the new widget in place, rendering its contents.

Click 'manage widgets' again and click the edit button next to a widget. You should see a dialog which shows the fields for that widget plus a field for the widget title. If you click cancel no changes should be shown in the widget list. If you click 'Edit widget' it should update the widget list with your changes, but nothing should be saved to the server yet until you click 'Done'.

You should be able to delete or move widgets around without any problems while editing or adding widgets.

#### Screenshots
Click "+ Add widget", then clicking next before selecting a widget type:
select_widget_1

Clicking next before filling out the widget title:
screenshot from 2018-12-21 16-54-08

Edit widget:
screenshot from 2018-12-21 16-54-49

No mobile screenshots because #1518 is not done

### CODE REVIEW

**Comment by aliceriot** [comment]
The UI lets me make a URL widget and it seems to save it, but when I refresh the page it doesn't show up in the UI. maybe I have something misconfigured? other than that the functionality seems to work well.

**Comment by aliceriot** [comment]
My only worry is that, with the separate dialog opening, it isn't clear enough to the user that their changes aren't going to be saved when they confirm the dialog box. this is more of a UI / UX type question than a code question though (@pdpinch).

File: static/js/components/widgets/WidgetEditDialog.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+          options={specs.map(spec => ({
+            label: spec.widget_type,
+            value: spec.widget_type
+          }))}
+        />
+      </React.Fragment>
+    )
+  }
+
+  renderConfiguration = () => {
```
[the remark this answers is not in the document]
Comment (noisecapella) [reply static/js/components/widgets/WidgetEditDialog.js]: :heavy_check_mark:

File: static/js/components/widgets/WidgetEditDialog.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    const { specs, dialogData } = this.props
+
+    const validation = dialogData ? dialogData.validation : {}
+    const spec = specs.filter(
+      spec => spec.widget_type === this.getValue(widgetTypeLens)
+    )[0]
+
+    return (
+      <React.Fragment>
+        {validationMessage(validation.title)}
```
[the remark this answers is not in the document]
Comment (noisecapella) [reply static/js/components/widgets/WidgetEditDialog.js]: :heavy_check_mark:

File: static/js/components/widgets/WidgetField.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import type { WidgetFieldSpec } from "../../flow/widgetTypes"
+
+type Props = {
+  fieldSpec: WidgetFieldSpec,
+  onChange: (event: any) => void,
+  value: any
+}
+
+const WidgetField = ({ fieldSpec, value, onChange }: Props) => {
+  const valueOrDefault = value !== undefined ? value : fieldSpec.props.default
```
[the remark this answers is not in the document]
Comment (noisecapella) [reply static/js/components/widgets/WidgetField.js]: :heavy_check_mark:

File: static/js/actions/ui.js [hunk 4]
```diff
@@ -5,6 +5,11 @@ export const DIALOG_REMOVE_POST = "DIALOG_REMOVE_POST"
 export const DIALOG_REMOVE_COMMENT = "DIALOG_REMOVE_COMMENT"
 export const DIALOG_REMOVE_MEMBER = "DIALOG_REMOVE_MEMBER"
 
+export const DIALOG_EDIT_WIDGET = "DIALOG_EDIT_WIDGET"
+export const DIALOG_EDIT_WIDGET_SELECT_TYPE = "DIALOG_EDIT_WIDGET_SELECT_TYPE"
+export const DIALOG_EDIT_WIDGET_CONFIGURATION =
```
[the remark this answers is not in the document]
Comment (noisecapella) [reply static/js/actions/ui.js]: Sure, I'll look into this

[the remark this answers is not in the document]
Comment (noisecapella) [reply static/js/actions/ui.js]: :heavy_check_mark:

**Comment by noisecapella** [comment]
Any other comments?

**Comment by aliceriot** [comment]
I think this is totally unrelated to the changes in this PR, but I just noticed that clicking the 'delete' icon on any single widget in the 'editing' mode will delete all widgets which haven't yet been saved to the backend. It looks like this is happening because the widgets which haven't been saved yet do not have an `.id` property yet, and the `deleteInstance` method on the `WidgetListContainer` uses that prop to figure out which widgets to delete (in the case of not-yet-saved ones the value of this property will be `undefined`, which will then match every other not-yet-saved widget).

does it make sense to take care of that little bug in this PR, or should we file an issue?

**Comment by noisecapella** [comment]
@aliceriot Ah good catch. I'll make an issue for it

**Pull request merged**